File: packages/analytics/src/__tests__/AnalyticsModule.test.ts
import { describe, it, expect, beforeEach } from 'vitest';
import { RynBridge, MockTransport } from '@rynbridge/core';
import type { BridgeResponse } from '@rynbridge/core';
import { AnalyticsModule } from '../AnalyticsModule.js';

describe('AnalyticsModule', () => {
  let transport: MockTransport;
  let bridge: RynBridge;
  let analytics: AnalyticsModule;

  beforeEach(() => {
    transport = new MockTransport();
    bridge = new RynBridge({ timeout: 5000 }, transport);
    analytics = new AnalyticsModule(bridge);
  });

  function respondSuccess(payload: Record<string, unknown> = {}) {
    const sent = JSON.parse(transport.sent[transport.sent.length - 1]);
    const response: BridgeResponse = {
      id: sent.id,
      status: 'success',
      payload,
      error: null,
    };
    transport.simulateIncoming(JSON.stringify(response));
  }

  describe('logEvent', () => {
    it('sends fire-and-forget with name only', () => {
      analytics.logEvent({ name: 'page_view' });

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('logEvent');
      expect(sent.payload).toEqual({ name: 'page_view' });
    });

    it('sends fire-and-forget with name and params', () => {
      analytics.logEvent({ name: 'purchase', params: { amount: 100, currency: 'USD' } });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('logEvent');
      expect(sent.payload).toEqual({ name: 'purchase', params: { amount: 100, currency: 'USD' } });
    });
  });

  describe('setUserProperty', () => {
    it('sends fire-and-forget with key and value', () => {
      analytics.setUserProperty({ key: 'plan', value: 'premium' });

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('setUserProperty');
      expect(sent.payload).toEqual({ key: 'plan', value: 'premium' });
    });
  });

  describe('setUserId', () => {
    it('sends fire-and-forget with userId', () => {
      analytics.setUserId({ userId: 'user-123' });

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('setUserId');
      expect(sent.payload).toEqual({ userId: 'user-123' });
    });
  });

  describe('setScreen', () => {
    it('sends fire-and-forget with screen name', () => {
      analytics.setScreen({ name: 'HomeScreen' });

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('setScreen');
      expect(sent.payload).toEqual({ name: 'HomeScreen' });
    });
  });

  describe('resetUser', () => {
    it('sends fire-and-forget with empty payload', () => {
      analytics.resetUser();

      expect(transport.sent).toHaveLength(1);
      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('resetUser');
      expect(sent.payload).toEqual({});
    });
  });

  describe('setEnabled', () => {
    it('sends request and returns enabled state', async () => {
      const promise = analytics.setEnabled({ enabled: false });

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('setEnabled');
      expect(sent.payload).toEqual({ enabled: false });

      respondSuccess({ enabled: false });

      const result = await promise;
      expect(result).toEqual({ enabled: false });
    });

    it('can enable tracking', async () => {
      const promise = analytics.setEnabled({ enabled: true });
      respondSuccess({ enabled: true });

      const result = await promise;
      expect(result).toEqual({ enabled: true });
    });
  });

  describe('isEnabled', () => {
    it('sends request and returns current enabled state', async () => {
      const promise = analytics.isEnabled();

      const sent = JSON.parse(transport.sent[0]);
      expect(sent.module).toBe('analytics');
      expect(sent.action).toBe('isEnabled');
      expect(sent.payload).toEqual({});

      respondSuccess({ enabled: true });

      const result = await promise;
      expect(result).toEqual({ enabled: true });
    });
  });

  describe('error handling', () => {
    it('propagates bridge errors on setEnabled', async () => {
      const promise = analytics.setEnabled({ enabled: false });

      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Analytics provider not configured' },
      };
      transport.simulateIncoming(JSON.stringify(response));

      await expect(promise).rejects.toThrow('Analytics provider not configured');
    });

    it('propagates bridge errors on isEnabled', async () => {
      const promise = analytics.isEnabled();

      const sent = JSON.parse(transport.sent[0]);
      const response: BridgeResponse = {
        id: sent.id,
        status: 'error',
        payload: {},
        error: { code: 'UNKNOWN', message: 'Analytics provider not configured' },
      };
      transport.simulateIncoming(JSON.stringify(response));

      await expect(promise).rejects.toThrow('Analytics provider not configured');
    });
  });
});

File: android/analytics/src/test/kotlin/io/rynbridge/analytics/AnalyticsModuleTest.kt
package io.rynbridge.analytics

import io.rynbridge.core.*
import kotlinx.coroutines.test.runTest
import org.junit.jupiter.api.Assertions.*
import org.junit.jupiter.api.Test

class AnalyticsModuleTest {

    @Test
    fun `logEvent calls provider with name and params`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["logEvent"]!!

        val result = handler(mapOf(
            "name" to BridgeValue.string("button_click"),
            "params" to BridgeValue.dict(mapOf(
                "screen" to BridgeValue.string("home"),
                "count" to BridgeValue.int(5)
            ))
        ))
        assertTrue(result.isEmpty())
        assertEquals("button_click", provider.lastEventName)
        assertEquals("home", provider.lastEventParams?.get("screen")?.stringValue)
    }

    @Test
    fun `setUserProperty calls provider`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["setUserProperty"]!!

        val result = handler(mapOf(
            "key" to BridgeValue.string("plan"),
            "value" to BridgeValue.string("premium")
        ))
        assertTrue(result.isEmpty())
        assertEquals("plan", provider.lastPropertyKey)
        assertEquals("premium", provider.lastPropertyValue)
    }

    @Test
    fun `setUserId calls provider`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["setUserId"]!!

        val result = handler(mapOf("userId" to BridgeValue.string("user-123")))
        assertTrue(result.isEmpty())
        assertEquals("user-123", provider.lastUserId)
    }

    @Test
    fun `setScreen calls provider`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["setScreen"]!!

        val result = handler(mapOf("name" to BridgeValue.string("home")))
        assertTrue(result.isEmpty())
        assertEquals("home", provider.lastScreenName)
    }

    @Test
    fun `resetUser calls provider`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["resetUser"]!!

        val result = handler(emptyMap())
        assertTrue(result.isEmpty())
        assertTrue(provider.resetUserCalled)
    }

    @Test
    fun `setEnabled returns enabled state`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["setEnabled"]!!

        val result = handler(mapOf("enabled" to BridgeValue.bool(true)))
        assertEquals(true, result["enabled"]?.boolValue)
        assertEquals(true, provider.enabled)
    }

    @Test
    fun `isEnabled returns current state`() = runTest {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        val handler = module.actions["isEnabled"]!!

        val result = handler(emptyMap())
        assertEquals(true, result["enabled"]?.boolValue)
    }

    @Test
    fun `module name and version`() {
        val provider = MockAnalyticsProvider()
        val module = AnalyticsModule(provider)
        assertEquals("analytics", module.name)
        assertEquals("0.1.0", module.version)
    }
}

private class MockAnalyticsProvider : AnalyticsProvider {
    var lastEventName: String? = null
    var lastEventParams: Map<String, BridgeValue>? = null
    var lastPropertyKey: String? = null
    var lastPropertyValue: String? = null
    var lastUserId: String? = null
    var lastScreenName: String? = null
    var resetUserCalled = false
    var enabled = true

    override fun logEvent(name: String, params: Map<String, BridgeValue>) {
        lastEventName = name
        lastEventParams = params
    }

    override fun setUserProperty(key: String, value: String) {
        lastPropertyKey = key
        lastPropertyValue = value
    }

    override fun setUserId(userId: String) {
        lastUserId = userId
    }

    override fun setScreen(name: String) {
        lastScreenName = name
    }

    override fun resetUser() {
        resetUserCalled = true
    }

    override suspend fun setEnabled(enabled: Boolean): Boolean {
        this.enabled = enabled
        return enabled
    }

    override suspend fun isEnabled(): Boolean = enabled
}
